### PR TITLE
[dhctl] Check dhctl version against current release channel version before bootstrap

### LINKED ISSUE #5073: [preflight-check] Check the compatibility of the dhctl installer versions and the Deckhouse version being installed
### Preflight Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/deckhouse/deckhouse/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://github.com/deckhouse/deckhouse/issues) for an issue that matches the one I want to file, without success.

### Use case. Why is this important?

Preventing failures before bootstrap execution

### Proposed Solution

Before performing bootstrap, check the compatibility of the dhctl installer versions and the Deckhouse version being installed

### Additional Information

* use version Deckhouse from the file /deckhouse/version for deployment. May be, use the golang compile time constants for setting deckhouse version ```go build -ldflags "-X config.Version=1.0.0" ...```
* additionally, we can check the hash sum SHA256 of the Deckhouse image
* additionally, we may need to move the releaseChannel installation from configOverrides to the last bootstrap step.

### PR DESCRIPTION
**Pull request by mvasl**
## Description

Added a preflight check that ensures installer image is "compatible" with current Deckhouse version in selected release channel. If this preflight check fails, installation is aborted with an error message that describes the problem and provides means to fix it.

There is an override to this behavoiur that allows to force installation even if `dhctl` is not "compatible" via `DHCTL_CLI_PREFLIGHT_SKIP_INCOMPATIBLE_VERSION_CHECK` enviroment variable.

Term "compatible" here means that deckhouse container image digests match between dhctl build and deckhouse release channel image it is configured to install.

## Why do we need it, and what problem does it solve?

Closes #5073 and #3251 

## Manual testing
https://github.com/deckhouse/deckhouse/pull/5832#issuecomment-1728973475

## Checklist
- [x] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: dhctl
type: feature
summary: dhctl will check if it is compatible with deckhouse version it is bootstrapping
impact_level: default
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've started the workflow for label 'e2e/run/yandex-cloud'...

File: dhctl/pkg/preflight/preflight.go [hunk 1]
```diff
@@ -16,17 +16,24 @@ package preflight
...
 }
 
-func NewPreflightCheck(sshClient *ssh.Client) PreflightCheck {
+func NewPreflightCheck(sshClient *ssh.Client, config *deckhouse.Config) PreflightCheck {
 	return PreflightCheck{
-		sshClient: sshClient,
+		sshClient:               sshClient,
+		installConfig:           config,
+		imageDescriptorProvider: remoteDescriptorProvider{},
+		buildDigestProvider:     &dhctlBuildDigestProvider{DigestFilePath: app.DeckhouseImageDigestFile},
```
Comment (aleksey-su) [inline dhctl/pkg/preflight/preflight.go:36]: Maybe it's worth creating constructor functions that will return the corresponding interfaces as a type?

**Review by aleksey-su — CHANGES_REQUESTED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/static'...

**Review by name212 — APPROVED** [review]

**Review by name212 — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @name212. I've started the workflow for label 'e2e/run/static'...

**Review by name212 — DISMISSED** [review]

**Review by aleksey-su — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've started the workflow for label 'e2e/run/static'...

**Review by name212 — APPROVED** [review]

**Pull request merged**